lighthouse-core/audits/font-display.js

```diff
@@ -33,59 +36,91 @@ class FontDisplay extends Audit {
       title: str_(UIStrings.title),
       failureTitle: str_(UIStrings.failureTitle),
       description: str_(UIStrings.description),
-      requiredArtifacts: ['devtoolsLogs', 'Fonts'],
+      requiredArtifacts: ['devtoolsLogs', 'CSSUsage', 'URL'],
     };
   }
 
+  /**
+   *
+   * @param {LH.Artifacts} artifacts
+   */
+  static findPassingFontDisplayDeclarations(artifacts) {
+    /** @type {Set<string>} */
+    const passingURLs = new Set();
+
+    for (const stylesheet of artifacts.CSSUsage.stylesheets) {
+      const newlinesStripped = stylesheet.content.replace(/\n/g, ' ');
+      const fontFaceDeclarations = newlinesStripped.match(/@font-face\s*{(.*?)}/g) || [];
+      for (const declaration of fontFaceDeclarations) {
+        const rawFontDisplay = declaration.match(/font-display:(.*?);/);
+        if (!rawFontDisplay) continue;
+        const hasPassingFontDisplay = PASSING_FONT_DISPLAY_REGEX.test(rawFontDisplay[0]);
+        if (!hasPassingFontDisplay) continue;
+
+        const rawFontURLs = declaration.match(CSS_URL_GLOBAL_REGEX);
+        if (!rawFontURLs) continue;
+
+        const relativeURLs = rawFontURLs
+          // @ts-ignore - guaranteed to match from previous regex, pull URL group out
+          .map(s => s.match(CSS_URL_REGEX)[1].trim())
+          // remove any optional quotes before/after
+          .map(s => {
+            const firstChar = s.charAt(0);
+            if (firstChar === s.charAt(s.length - 1) && (firstChar === '"' || firstChar === '\'')) {
+              return s.substr(1, s.length - 2);
+            }
+
+            return s;
+          });
+
+        const absoluteURLs = relativeURLs.map(url => new URL(url, artifacts.URL.finalUrl));
```

Comment:
how do we want to deal with exceptions here? e.g. invalid font URL probably shouldn't throw the whole audit

Reply:
yeah good point, I'll filter on validity